File: backend/src/controllers/cryptoController.js
const coingeckoService = require('../services/coingeckoService');
const User = require('../models/User');
const Alert = require('../models/Alert');

// Get top cryptocurrencies
exports.getTopCoins = async (req, res) => {
  try {
    const { page = 1, perPage = 50, currency = 'usd' } = req.query;

    // Get user's preferred currency if authenticated
    let preferredCurrency = currency;
    if (req.user) {
      if (req.user.preferences && req.user.preferences.currency) {
        preferredCurrency = req.user.preferences.currency;
      }
    }

    const coins = await coingeckoService.getTopCoins(preferredCurrency, page, perPage);
    res.json({ coins, currency: preferredCurrency });
  } catch (error) {
    console.error('Get top coins error:', error);
    res.status(500).json({ message: 'Failed to fetch cryptocurrencies' });
  }
};

// Get coin details
exports.getCoinDetails = async (req, res) => {
  try {
    const { id } = req.params;
    const coin = await coingeckoService.getCoinDetails(id);
    res.json({ coin });
  } catch (error) {
    console.error('Get coin details error:', error);
    res.status(500).json({ message: 'Failed to fetch coin details' });
  }
};

// Get coin chart data
exports.getCoinChart = async (req, res) => {
  try {
    const { id } = req.params;
    const { days = 7, currency = 'usd' } = req.query;

    // Get user's preferred currency if authenticated
    let preferredCurrency = currency;
    if (req.user) {
      if (req.user.preferences && req.user.preferences.currency) {
        preferredCurrency = req.user.preferences.currency;
      }
    }

    const chartData = await coingeckoService.getCoinChart(id, preferredCurrency, days);
    res.json({ chartData, currency: preferredCurrency });
  } catch (error) {
    console.error('Get coin chart error:', error);
    res.status(500).json({ message: 'Failed to fetch chart data' });
  }
};

// Search coins
exports.searchCoins = async (req, res) => {
  try {
    const { query } = req.query;
    if (!query) {
      return res.status(400).json({ message: 'Search query is required' });
    }

    const results = await coingeckoService.searchCoins(query);
    res.json({ results });
  } catch (error) {
    console.error('Search coins error:', error);
    res.status(500).json({ message: 'Failed to search coins' });
  }
};

// Get trending coins
exports.getTrendingCoins = async (req, res) => {
  try {
    const trending = await coingeckoService.getTrendingCoins();
    res.json({ trending });
  } catch (error) {
    console.error('Get trending coins error:', error);
    res.status(500).json({ message: 'Failed to fetch trending coins' });
  }
};

// Get global market data
exports.getGlobalData = async (req, res) => {
  try {
    const { currency = 'usd' } = req.query;
    const globalData = await coingeckoService.getGlobalData(currency);
    res.json({ globalData });
  } catch (error) {
    console.error('Get global data error:', error);
    res.status(500).json({ message: 'Failed to fetch global market data' });
  }
};

// Get user's watchlist
exports.getWatchlist = async (req, res) => {
  try {
    console.log('getWatchlist: req.user is', req.user);
    const user = req.user; // User is already attached by auth middleware
    if (!user.watchlist || user.watchlist.length === 0) {
      return res.json({ watchlist: [] });
    }

    // Get detailed data for watchlist coins
    const coins = await coingeckoService.getCoinsPrices(user.watchlist);
    res.json({ watchlist: coins });
  } catch (error) {
    console.error('Get watchlist error:', error);
    res.status(500).json({ message: 'Failed to fetch watchlist' });
  }
};

// Add coin to watchlist
exports.addToWatchlist = async (req, res) => {
  try {
    const { coinId } = req.body;
    if (!coinId) {
      return res.status(400).json({ message: 'Coin ID is required' });
    }

    const user = req.user;
    if (!user) {
      console.error('addToWatchlist: req.user is missing!');
      return res.status(401).json({ message: 'User not authenticated context missing' });
    }
    if (!user.watchlist) {
      // Initialize if missing (defensive)
      user.watchlist = [];
    }
    if (user.watchlist.includes(coinId)) {
      return res.status(400).json({ message: 'Coin already in watchlist' });
    }

    user.watchlist.push(coinId);
    await user.save();

    // Get updated watchlist with coin details
    const coins = await coingeckoService.getCoinsPrices(user.watchlist);
    res.json({ message: 'Added to watchlist', watchlist: coins });
  } catch (error) {
    console.error('Add to watchlist error:', error);
    res.status(500).json({ message: 'Failed to add to watchlist' });
  }
};

// Remove coin from watchlist
exports.removeFromWatchlist = async (req, res) => {
  try {
    const { coinId } = req.params;

    const user = req.user;
    user.watchlist = user.watchlist.filter(id => id !== coinId);
    await user.save();

    // Get updated watchlist with coin details
    const coins = await coingeckoService.getCoinsPrices(user.watchlist);
    res.json({ message: 'Removed from watchlist', watchlist: coins });
  } catch (error) {
    console.error('Remove from watchlist error:', error);
    res.status(500).json({ message: 'Failed to remove from watchlist' });
  }
};

// Create alert
exports.createAlert = async (req, res) => {
  try {
    console.log('createAlert body:', req.body);
    const { symbol, targetPrice, direction } = req.body;

    if (!symbol || !direction || !targetPrice) {
      return res.status(400).json({ message: 'Missing required fields: symbol, direction, targetPrice' });
    }

    const alert = new Alert({
      user: req.user._id,
      symbol,
      targetPrice,
      direction,
    });

    await alert.save();

    // Add alert to user's alerts array
    const user = req.user;
    user.alerts.push(alert._id);
    await user.save();

    res.status(201).json({ message: 'Alert created successfully', alert });
  } catch (error) {
    console.error('Create alert error:', error);
    res.status(500).json({ message: 'Failed to create alert' });
  }
};

// Get user's alerts
exports.getAlerts = async (req, res) => {
  try {
    const alerts = await Alert.find({ user: req.user._id })
      .sort({ createdAt: -1 });
    res.json({ alerts });
  } catch (error) {
    console.error('Get alerts error:', error);
    res.status(500).json({ message: 'Failed to fetch alerts' });
  }
};

// Delete alert
exports.deleteAlert = async (req, res) => {
  try {
    const { alertId } = req.params;

    const alert = await Alert.findOneAndDelete({
      _id: alertId,
      user: req.user._id
    });

    if (!alert) {
      return res.status(404).json({ message: 'Alert not found' });
    }

    // Remove alert from user's alerts array
    const user = req.user;
    user.alerts = user.alerts.filter(id => id.toString() !== alertId);
    await user.save();

    res.json({ message: 'Alert deleted successfully' });
  } catch (error) {
    console.error('Delete alert error:', error);
    res.status(500).json({ message: 'Failed to delete alert' });
  }
};

File: backend2/src/main/java/com/coinvista/backend/controller/CryptoController.java
package com.coinvista.backend.controller;

import com.coinvista.backend.dto.AlertDto;
import com.coinvista.backend.model.Alert;
import com.coinvista.backend.model.User;
import com.coinvista.backend.service.AlertService;
import com.coinvista.backend.service.CoinGeckoService;
import com.coinvista.backend.service.WatchlistService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/crypto")
@RequiredArgsConstructor
public class CryptoController {

    private final CoinGeckoService coinGeckoService;
    private final WatchlistService watchlistService;
    private final AlertService alertService;

    // ── Public endpoints ──────────────────────────────────────────────────────

    @GetMapping("/coins/top")
    public ResponseEntity<?> getTopCoins(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "50") int perPage,
            @RequestParam(defaultValue = "usd") String currency) {
        try {
            Object coins = coinGeckoService.getTopCoins(currency, page, perPage);
            return ResponseEntity.ok(Map.of("coins", coins, "currency", currency));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to fetch cryptocurrencies"));
        }
    }

    @GetMapping("/coins/{id}")
    public ResponseEntity<?> getCoinDetails(@PathVariable String id) {
        try {
            Object coin = coinGeckoService.getCoinDetails(id);
            return ResponseEntity.ok(Map.of("coin", coin));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to fetch coin details"));
        }
    }

    @GetMapping("/coins/{id}/chart")
    public ResponseEntity<?> getCoinChart(
            @PathVariable String id,
            @RequestParam(defaultValue = "7") int days,
            @RequestParam(defaultValue = "usd") String currency) {
        try {
            Object chartData = coinGeckoService.getCoinChart(id, currency, days);
            return ResponseEntity.ok(Map.of("chartData", chartData, "currency", currency));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to fetch chart data"));
        }
    }

    @GetMapping("/search")
    public ResponseEntity<?> searchCoins(@RequestParam String query) {
        try {
            Object results = coinGeckoService.searchCoins(query);
            return ResponseEntity.ok(Map.of("results", results));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to search coins"));
        }
    }

    @GetMapping("/trending")
    public ResponseEntity<?> getTrendingCoins() {
        try {
            Object trending = coinGeckoService.getTrendingCoins();
            return ResponseEntity.ok(Map.of("trending", extractTrendingCoins(trending)));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to fetch trending coins"));
        }
    }

    @GetMapping("/global")
    public ResponseEntity<?> getGlobalData(@RequestParam(defaultValue = "usd") String currency) {
        try {
            Object globalData = coinGeckoService.getGlobalData();
            return ResponseEntity.ok(Map.of("globalData", globalData));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to fetch global market data"));
        }
    }

    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> extractTrendingCoins(Object trending) {
        if (trending instanceof Map<?, ?> map && map.get("coins") instanceof List<?> coins) {
            return coins.stream()
                    .filter(Map.class::isInstance)
                    .map(item -> (Map<String, Object>) item)
                    .toList();
        }

        if (trending instanceof List<?> coins) {
            return coins.stream()
                    .filter(Map.class::isInstance)
                    .map(item -> (Map<String, Object>) item)
                    .toList();
        }

        return List.of();
    }

    // ── Watchlist (protected) ─────────────────────────────────────────────────

    @GetMapping("/watchlist")
    public ResponseEntity<?> getWatchlist(@AuthenticationPrincipal User user) {
        try {
            Object watchlist = watchlistService.getWatchlist(user.getId());
            return ResponseEntity.ok(Map.of("watchlist", watchlist));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to fetch watchlist"));
        }
    }

    @PostMapping("/watchlist")
    public ResponseEntity<?> addToWatchlist(
            @AuthenticationPrincipal User user,
            @RequestBody Map<String, String> body) {
        try {
            String coinId = body.get("coinId");
            if (coinId == null || coinId.isBlank()) {
                return ResponseEntity.badRequest().body(Map.of("message", "Coin ID is required"));
            }
            Object watchlist = watchlistService.addToWatchlist(user.getId(), coinId);
            return ResponseEntity.ok(Map.of("message", "Added to watchlist", "watchlist", watchlist));
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(Map.of("message", e.getMessage()));
        }
    }

    @DeleteMapping("/watchlist/{coinId}")
    public ResponseEntity<?> removeFromWatchlist(
            @AuthenticationPrincipal User user,
            @PathVariable String coinId) {
        try {
            Object watchlist = watchlistService.removeFromWatchlist(user.getId(), coinId);
            return ResponseEntity.ok(Map.of("message", "Removed from watchlist", "watchlist", watchlist));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to remove from watchlist"));
        }
    }

    // ── Alerts (protected) ────────────────────────────────────────────────────

    @GetMapping("/alerts")
    public ResponseEntity<?> getAlerts(@AuthenticationPrincipal User user) {
        try {
            List<Alert> alerts = alertService.getAlerts(user.getId());
            return ResponseEntity.ok(Map.of("alerts", alerts));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to fetch alerts"));
        }
    }

    @PostMapping("/alerts")
    public ResponseEntity<?> createAlert(
            @AuthenticationPrincipal User user,
            @Valid @RequestBody AlertDto.CreateRequest request) {
        try {
            Alert alert = alertService.createAlert(user.getId(), request);
            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(Map.of("message", "Alert created successfully", "alert", alert));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to create alert"));
        }
    }

    @DeleteMapping("/alerts/{alertId}")
    public ResponseEntity<?> deleteAlert(
            @AuthenticationPrincipal User user,
            @PathVariable String alertId) {
        try {
            alertService.deleteAlert(alertId, user.getId());
            return ResponseEntity.ok(Map.of("message", "Alert deleted successfully"));
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", e.getMessage()));
        }
    }
}
